File: parent/142 -- Linked List Cycle II/mine1.py
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Time: O(n). Slow version
        # Space: O(n)
        curr = head
        visit = set()
        while curr and curr not in visit:
            visit.add(curr)
            curr = curr.next
        
        if curr == None:
            return None
        return curr
        

File: parent/142 -- Linked List Cycle II/mine1.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func detectCycle(head *ListNode) *ListNode {
	visited := map[*ListNode]bool{}
	_, isVisited := visited[head]

	for head != nil && !isVisited {
		visited[head] = true
		head = head.Next
		_, isVisited = visited[head]

	}
	if head == nil {
		return nil
	}
	return head
}
